go/service/chat_local.go

```diff
@@ -180,87 +180,170 @@ func (h *chatLocalHandler) CompleteAndCanonicalizeTlfName(ctx context.Context, t
-		resp, err := h.boxer.tlf.CryptKeys(ctx, criteria.TlfName)
+	// TODO: move these criterian to server using queries
+	appendMaybe := func(conv chat1.Conversation) (shouldContinue bool, err error) {
+		info, _, maxMessages, err := h.getConversationInfo(ctx, conv)
 		if err != nil {
-			return nil, err
+			return false, err
 		}
-		tlfIDb := resp.TlfID.ToBytes()
-		if tlfIDb == nil {
-			return nil, errors.New("invalid TLF ID acquired")
+		c := keybase1.ConversationLocal{
+			Id:       info.Id,
+			Info:     &info,
+			Messages: maxMessages,
 		}
-		conversationsRemote, err := h.remoteClient().GetInboxByTLFIDRemote(ctx, tlfIDb)
-		if err != nil {
-			return nil, err
+
+		if topicTypes != nil && !topicTypes[info.TopicType] {
+			return true, nil
 		}
-		for _, cr := range conversationsRemote.Convs {
-			info, _, err := h.getConversationInfo(ctx, cr)
-			if err != nil {
-				return nil, err
+
+		switch {
+		case !since.IsZero(): // since is present; limit is ignored
+			var newest time.Time
+			for _, m := range maxMessages {
+				t := gregor1.FromTime(m.ServerHeader.Ctime)
+				if t.After(newest) {
+					newest = t
+				}
 			}
-			if info.TlfName != criteria.TlfName {
-				// check again using signed information to make sure it's the correct
-				// conversation
-				return nil, libkb.UnexpectedChatDataFromServer{Msg: fmt.Sprintf("Unexpected data is returned from server. We asked for %v, but got conversation for %v. TODO: handle tlfName changes properly for SBS case", criteria.TlfName, info.TlfName)}
+
+			if newest.Before(since2) { // too old; don't care anymore
+				return false, nil
+			} else if newest.Before(since) { // not in interesting range; but caller might be interested in this in the future, so put in res.More
+				res.More = append(res.More, c)
+				return true, nil
+			} else { // caller is interested in this conversation
+				res.Conversations = append(res.Conversations, c)
+				return true, nil
 			}
-			appendMaybe(info)
+		case arg.Limit > 0: // since is not present; try to use limit
+			if len(res.Conversations) < arg.Limit { // caller is interested in this conversation
+				res.Conversations = append(res.Conversations, c)
+				return true, nil
+			} else if len(res.More) < arg.Limit { // not in interesting range; but caller might be interested in this in the future, so put in res.More
+				res.More = append(res.More, c)
+				return true, nil
+			} else { // too much; don't care anymore
+				return false, nil
+			}
+		default: // not limiting number of items in result at all
+			res.Conversations = append(res.Conversations, c)
+			return true, nil
 		}
-	} else { // slow case: tlfName not specified; need to search through inbox
-		ipagination := &chat1.Pagination{Num: 20}
-	getinbox:
-		for i := 0; i < 10000; /* in case we have a server bug */ i++ {
-			iview, err := h.GetInboxLocal(ctx, ipagination)
+
+	}
+
+	ipagination := &chat1.Pagination{Num: 20}
+	if arg.Limit != 0 {
+		ipagination.Num = arg.Limit * 2
```

Comment:
The idea is that there might be some conversation in the returned page not meeting the criteria, making it necessary to fetch another page. After we have server side query support for these criteria, it should become cleaner